core/src/main/java/org/apache/iceberg/avro/AvroIO.java

```diff
@@ -131,4 +144,55 @@ public boolean markSupported() {
       return stream.markSupported();
     }
   }
+
+  static long findStartingRowPos(Supplier<SeekableInputStream> open, long start) {
+    try (SeekableInputStream in = open.get()) {
+      // use a direct decoder that will not buffer so the position of the input stream is accurate
+      BinaryDecoder decoder = DecoderFactory.get().directBinaryDecoder(in, null);
+
+      // an Avro file's layout looks like this:
+      //   header|block|block|...
+      // the header contains:
+      //   magic|string-map|sync
+      // each block consists of:
+      //   row-count|compressed-size-in-bytes|block-bytes|sync
+
+      // it is necessary to read the header here because this is the only way to get the expected file sync bytes
+      byte[] magic = MAGIC_READER.read(decoder, null);
+      if (!Arrays.equals(AVRO_MAGIC, magic)) {
+        throw new InvalidAvroMagicException("Not an Avro file");
+      }
+
+      META_READER.read(decoder, null); // ignore the file metadata, it isn't needed
+      byte[] fileSync = SYNC_READER.read(decoder, null);
+
+      // the while loop reads row counts and seeks past the block bytes until the next sync pos is >= start, which
+      // indicates that the next sync is the start of the split.
+      byte[] blockSync = new byte[16];
+      long totalRows = 0;
+      long nextSyncPos = in.getPos();
+
+      while (nextSyncPos < start) {
```

Comment:
Not sure if this valid start state, but I think there can be an edge case here
```
row-count|compressed-size-in-bytes|block-bytes|sync|EOF
                                                 ^
                                                 |
                                               start
```
In this case it will seek and read the sync and then continue to read next block row count even after EOF

Reply:
Yeah, this is a bit strange in Avro. If the `start` of a split is past the start of the sync, then the split will actually start from the next sync.

To find the actual start of the split, not the start of the assigned range, Avro will read 16 bytes from `start` and then continually read one more byte and see if the rolling 16 match the sync. In the case that `start` is `syncPos + 1`, it won't match the sync bytes until the _next_ sync, even though the actual block starts after `start`.

Reply:
The loop condition is correct with respect to `start` because a block starts at the beginning of a sync. You're right about the EOF behavior, though. I'll take a look at that.

Reply:
Thanks for the insightful comment! (before it went away 😃 ) I was not familiar Avro's split reading behavior.

Reply:
Sorry, I should have edited, not deleted. I missed the part about it being just before the EOF.

Reply:
The fix is to catch `EOFException` and return the number of rows.